File: vito/litho-slideshow/src/main/java/com/facebook/fresco/vito/litho/slideshow/FrescoVitoSlideshowDrawable.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.litho.slideshow;

import android.graphics.drawable.Drawable;
import com.facebook.common.internal.Preconditions;
import com.facebook.drawee.drawable.FadeDrawable;
import com.facebook.fresco.vito.core.FrescoDrawable2;
import com.facebook.infer.annotation.Nullsafe;
import java.util.TimerTask;
import javax.annotation.Nullable;

@Nullsafe(Nullsafe.Mode.LOCAL)
public class FrescoVitoSlideshowDrawable extends FadeDrawable {

  private int mCurrentLayer = 0;
  private @Nullable TimerTask mTimerTask;

  public FrescoVitoSlideshowDrawable(
      FrescoDrawable2 drawable1, FrescoDrawable2 drawable2, FrescoDrawable2 drawable3) {
    super(new Drawable[] {drawable1, drawable2, drawable3});
  }

  public void setTimerTask(TimerTask timerTask) {
    mTimerTask = timerTask;
  }

  @Nullable
  public TimerTask getTimerTask() {
    return mTimerTask;
  }

  public FrescoDrawable2 getCurrent() {
    return Preconditions.checkNotNull((FrescoDrawable2) getDrawable(mCurrentLayer));
  }

  public FrescoDrawable2 getNext() {
    return Preconditions.checkNotNull((FrescoDrawable2) getDrawable(getNextLayerIndex()));
  }

  public FrescoDrawable2 getPrevious() {
    return Preconditions.checkNotNull((FrescoDrawable2) getDrawable(getPreviousLayerIndex()));
  }

  public void fadeToNext() {
    final int prev = getPreviousLayerIndex();
    final int next = getNextLayerIndex();

    beginBatchMode();
    fadeUpToLayer(next);
    hideLayerImmediately(prev);
    endBatchMode();
    mCurrentLayer = next;
  }

  @Override
  public void reset() {
    if (mTimerTask != null) {
      mTimerTask.cancel();
    }
    mTimerTask = null;
    super.reset();
    mCurrentLayer = 0;
  }

  private int getNextLayerIndex() {
    return (mCurrentLayer + 1) % getNumberOfLayers();
  }

  private int getPreviousLayerIndex() {
    return (mCurrentLayer - 1 + getNumberOfLayers()) % getNumberOfLayers();
  }
}

File: vito/litho-slideshow/src/main/java/com/facebook/fresco/vito/litho/slideshow/FrescoVitoSlideshowDrawable.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.litho.slideshow

import android.graphics.drawable.Drawable
import com.facebook.drawee.drawable.FadeDrawable
import com.facebook.fresco.vito.core.FrescoDrawableInterface
import java.util.TimerTask

class FrescoVitoSlideshowDrawable<T>(drawable1: T, drawable2: T, drawable3: T) :
    FadeDrawable(arrayOf<Drawable>(drawable1, drawable2, drawable3)) where
T : Drawable,
T : FrescoDrawableInterface {

  private var currentLayer = 0
  var timerTask: TimerTask? = null
  val currentImage: FrescoDrawableInterface
    get() = checkNotNull(getDrawable(currentLayer) as FrescoDrawableInterface?)

  val nextImage: FrescoDrawableInterface
    get() = checkNotNull(getDrawable(nextLayerIndex) as FrescoDrawableInterface?)

  val previousImage: FrescoDrawableInterface
    get() = checkNotNull(getDrawable(previousLayerIndex) as FrescoDrawableInterface?)

  private val nextLayerIndex: Int
    get() = (currentLayer + 1) % numberOfLayers

  private val previousLayerIndex: Int
    get() = (currentLayer - 1 + numberOfLayers) % numberOfLayers

  fun fadeToNext() {
    val prev = previousLayerIndex
    val next = nextLayerIndex
    beginBatchMode()
    fadeUpToLayer(next)
    hideLayerImmediately(prev)
    endBatchMode()
    currentLayer = next
  }

  override fun reset() {
    timerTask?.cancel()
    timerTask = null
    super.reset()
    currentLayer = 0
  }
}
